programs/sunrise_beam/src/lib.rs

```diff
@@ -2,88 +2,239 @@
     pub fn remove_beam(ctx: Context<RemoveBeam>, beam: Pubkey) -> Result<()> {
         remove_beam::handler(ctx, beam)
     }
-}
 
-pub fn check_beam_parameters(
-    state: &ControllerState,
-    beam: &AccountInfo,
-    cpi_program_id: &Pubkey,
-) -> Result<()> {
-    if !state.contains_beam(beam.key) {
-        return Err(BeamProgramError::BeamNotPresent.into());
-    }
-    if beam.owner != cpi_program_id {
-        return Err(BeamProgramError::UnexpectedCallingProgram.into());
+    /// Change gsol mint authority to a new account.
+    pub fn export_mint_authority(ctx: Context<ExportMintAuthority>) -> Result<()> {
+        export_mint_authority::handler(ctx)
     }
+}
+
+#[derive(Accounts)]
+#[instruction(input: RegisterStateInput)]
+pub struct RegisterState<'info> {
+    #[account(mut)]
+    pub payer: Signer<'info>,
+    #[account(
+        init,
+        payer = payer,
+        space = State::size(input.initial_capacity),
+    )]
+    pub state: Account<'info, State>,
+    pub gsol_mint: Account<'info, Mint>,
+    /// CHECK: a PDA of this Sunrise program.
+    #[account(seeds = [state.key().as_ref(), GSOL_AUTHORITY], bump)]
+    pub gsol_mint_authority: UncheckedAccount<'info>,
+    pub system_program: Program<'info, System>,
+    pub token_program: Program<'info, Token>,
+    pub rent: Sysvar<'info, Rent>,
+}
+
+#[derive(Accounts)]
+#[instruction(state: Pubkey)]
+pub struct RegisterBeam<'info> {
+    #[account(
+        mut,
+        has_one = update_authority
+    )]
+    pub state: Account<'info, State>,
+    #[account(mut)]
+    pub payer: Signer<'info>,
+    pub update_authority: Signer<'info>,
+    /// CHECK: The state account of the program to be registered.
+    #[account(constraint = beam_account.key() == state.key())]
+    pub beam_account: UncheckedAccount<'info>,
+    pub system_program: Program<'info, System>,
+}
+
+#[derive(Accounts)]
+pub struct UpdateBeamAllocations<'info> {
+    #[account(
+        mut,
+        has_one = update_authority
+    )]
+    pub state: Account<'info, State>,
+    pub update_authority: Signer<'info>,
+}
+
+#[derive(Accounts)]
+pub struct RemoveBeam<'info> {
+    #[account(
+        mut,
+        has_one = update_authority
+    )]
+    pub state: Account<'info, State>,
+    pub update_authority: Signer<'info>,
+}
+
+#[derive(Accounts)]
+pub struct BurnGsol<'info> {
+    #[account(
+        mut,
+        has_one = gsol_mint,
+    )]
+    pub state: Box<Account<'info, State>>,
+    pub beam: Signer<'info>,
+    #[account(mut)]
+    pub gsol_mint: Box<Account<'info, Mint>>,
+    pub mint_gsol_to_owner: Signer<'info>,
+    #[account(
+        mut,
+        token::mint = gsol_mint,
+        token::authority = mint_gsol_to_owner
+    )]
+    pub mint_gsol_to: Box<Account<'info, TokenAccount>>,
+    /// CHECK: We check that it's the expected ID.
+    #[account(constraint = SysvarInstructions::check_id(sysvar.key))]
+    pub sysvar: UncheckedAccount<'info>,
```

Comment:
which sysvar is this?

Reply:
It's a dummy type [here](https://docs.rs/solana-program/latest/solana_program/sysvar/instructions/struct.Instructions.html) that implements the `Sysvar` trait to access the ID of the instructions sysvar. I just made a change to use the more direct [sysvar::instructions::ID](https://docs.rs/solana-program/latest/solana_program/sysvar/instructions/static.ID.html) instead.